### DISCUSSION TITLE
식별자 네이밍 규칙은 어떤 것들이 있나요?

### DISCUSSION THREAD

**Discussion by publdaze** [publdaze/js-study#17] [변수]
- 식별자 네이밍 규칙은 어떤 것들이 있나요?
  - 네이밍 컨벤션의 종류에는 어떤 것이 있는 지와 각 종류에 대해서 설명해주세요.
  - 네이밍 컨벤션이 아닌 네이밍 규칙에는 어떤 게 있는 지 설명해주세요.
    - 예약어 중 let도 식별자로 사용이 불가능 한가요?
      - strict mode에 대해 설명해주세요. 
  - 프로젝트에서 무엇을 자주 사용했나요? 그 이유는 무엇인가요?

**Comment by pipisebastian** [comment]
> 식별자 네이밍 규칙은 어떤 것들이 있나요?
>네이밍 컨벤션의 종류에는 어떤 것이 있는 지와 각 종류에 대해서 설명해주세요.

카멜 케이스(camelCase), 스네이크 케이스(snake_case), 파스칼 케이스(PacalCase), 헝가리안 케이스(strName - 변수앞에 타입까지), 케밥 케이스(kebab-case)

> 네이밍 컨벤션이 아닌 네이밍 규칙에는 어떤 게 있는 지 설명해주세요

- 식별자는 특수문자를 제외한 문자, 언더스코어(_), 달러 기호($)로 시작해야 한다. (숫자로 시작 X) 
- 예약어는 식별자로 사용할 수 없다. 

>예약어 중 let도 식별자로 사용이 불가능 한가요? strict mode에 대해 설명해주세요.

let은 식별자로 가능합니다. 그러나 strict mode에서는 불가능합니다.
strict mode란 자바스크립트에서 엄격한 오류 검사를 적용해 줍니다. 이는  "use strict" 로 최상단에 선언하여 사용가능합니다

>프로젝트에서 무엇을 자주 사용했나요? 그 이유는 무엇인가요?

목적에 따라 다른데, 일반적으로 함수나 변수의 이름은 카멜케이스, 생성자 함수나 클래스에는 파스칼 케이스를 사용합니다. 상수는 UPPER_SNAK_CASE를 사용했습니다. 
팀프로젝트에서 보편적으로 정해져있는 규칙을 따름으로써, 코드의 가독성과 효율성이 향상됩니다.

**Comment by publdaze** [comment]
> - 식별자 네이밍 규칙은 어떤 것들이 있나요?
> - 네이밍 컨벤션의 종류에는 어떤 것이 있는 지와 각 종류에 대해서 설명해주세요.
- camelCase - 식별자의 첫글자 소문자, 각 단어 구분 시 단어 첫 글자 대문자
- PascalCase - 식별자의 첫글자 대문자, 각 단어 구분 시 단어 첫 글자 대문자
- snake_case - 소문자로 구성, 각 단어 구분 시 단어 사이 `_`
- UPPER_SNAKE_CASE - 대문자로 구성, 각 단어 구분 시 단어 사이 `_`
> - 프로젝트에서 무엇을 자주 사용했나요? 그 이유는 무엇인가요?
- 일반적인 변수나 함수에는 camelCase 사용, 클래스에는 PascalCase 사용 -> 일반적으로 많이 사용하는 컨벤션 + ECMAScript 사양에 정의된 컨벤션 -> 가독성 상승과 다른 개발자의 원활한 커뮤니케이션 위함
- 리액트 프로젝트에서 컴포넌트 PascalCase -> 공식문서에 컴포넌트는 대문자로 사용하라는 규칙 명시하고 있음
- 상수 정의 UPPER_SNAKE_CASE -> 다른 변수와 구분을 위해
> - 네이밍 컨벤션이 아닌 네이밍 규칙에는 어떤 게 있는 지 설명해주세요.
- 예약어는 식별자로 사용 X
- $, _가 아닌 특수 문자 사용 X
- 숫자로 시작 X
- 대소문자 구분
> - 예약어 중 let도 식별자로 사용이 불가능 한가요?
- let은 식별자 가능
- 아직 예약어로 등록되지 않았지만 strict 모드에서는 예약어로 취급하여 식별자 불가능
> - strict mode에 대해 설명해주세요. 
- 자바스크립트 문법을 좀 더 엄격하게 적용하여 오류 발생 가능성이 있는 코드에 명시적인 에러를 발생시키는 모드
